File: codility/lessons/7_stack_and_queue/stone_wall.py
# https://app.codility.com/programmers/lessons/7-stacks_and_queues/stone_wall/

def solution(H):
    height_stack = []
    blocks_used = 0

    for wall_height in H:

        while height_stack and height_stack[-1] > wall_height:
            height_stack.pop()

        if not height_stack or height_stack[-1] < wall_height:
            height_stack.append(wall_height)
            blocks_used += 1

    return blocks_used

File: codility/lessons/7_stack_and_queue/stone_wall.cpp
#include <stack>
#include <vector>
using namespace std;

int solution(vector<int> &heights) {
    int blocks_used = 0;
    stack<int> blocks;

    for (int h : heights) {
        while (!blocks.empty() && blocks.top() > h) {
            blocks.pop();
        }
        if (blocks.empty() || h > blocks.top()) {
            blocks_used++;
            blocks.push(h);
        }
    }

    return blocks_used;
}
